sympy/functions/combinatorial/numbers.py

```diff
@@ -1861,9 +1861,26 @@ def stirling(n, k, d=None, kind=2, signed=False):
 def _nT(n, k):
     """Return the partitions of ``n`` items into ``k`` parts. This
     is used by ``nT`` for the case when ``n`` is an integer."""
+    if k > n or k < 0:
+        return 0
+    if k == n or k == 1:
+        return 1
     if k == 0:
-        return 1 if k == n else 0
-    return sum(_nT(n - k, j) for j in range(min(k, n - k) + 1))
+        return 0
+    if k == 2:
+        return n//2
+    d = n - k
+    if d in (1, 2):
```

Comment:
Could this be more efficient that `if d <= 2`? We know that `k < n` and therefore `d > 0`.

Reply:
I think its already there -> that is if d<=2, means d is in (1,2) and its already implemented.

Reply:
The question is, which one of the two equivalent tests is more efficient, `d in (1, 2)` or `d <= 2`.

Reply:
Okay, i think d<=2 is more efficient. I will change it and add more test cases.

Reply:
@jksuom  @smichr  I have made the changes. Are  there any more changes in this pr to be done.

Reply:
Could k < 0 be n*k < 0?

Reply:
@smichr  if i change k<0 to n*k<0 it will give wrong answer. e.g  when n=k=-2, then it will pass he first  condition ( e.g k>n or n*k<0)  then in 2nd condition since n=k it will return 1.

Reply:
The present conditions look pretty good to me. The main part of the algorithm (starting with `M = n`) presupposes that `0 < k <= n`. The first lines of the code must handle the remaining cases.

- `k <= n` is easy to guarantee:
    - if `k > n`, return 0.

- `k > 0` is split into two parts:
    -  if `k < 0`, return 0
    -  if `k == 0`, return 1 if `n == 0` else 0
       (The `n == 0` case is currently covered by the `k == n` test.)

The other commands are optimisations taking care of special cases (saving time but not necessary).